pkg/services/authn/clients/ext_jwt.go

```diff
@@ -0,0 +1,226 @@
+package clients
+
+import (
+	"context"
+	"fmt"
+	"net/http"
+	"strconv"
+	"strings"
+	"time"
+
+	"github.com/go-jose/go-jose/v3/jwt"
+	"golang.org/x/exp/slices"
+
+	"github.com/grafana/grafana/pkg/infra/log"
+	"github.com/grafana/grafana/pkg/services/authn"
+	"github.com/grafana/grafana/pkg/services/signingkeys"
+	"github.com/grafana/grafana/pkg/services/user"
+	"github.com/grafana/grafana/pkg/setting"
+)
+
+var _ authn.Client = new(ExtendedJWT)
+
+var (
+	acceptedSigningMethods = []string{"RS256", "ES256"}
+	timeNow                = time.Now
+)
+
+const (
+	rfc9068ShortMediaType = "at+jwt"
+	rfc9068MediaType      = "application/at+jwt"
+)
+
+func ProvideExtendedJWT(userService user.Service, cfg *setting.Cfg, signingKeys signingkeys.Service) *ExtendedJWT {
+	return &ExtendedJWT{
+		cfg:         cfg,
+		log:         log.New(authn.ClientExtendedJWT),
+		userService: userService,
+		signingKeys: signingKeys,
+	}
+}
+
+type ExtendedJWT struct {
+	cfg         *setting.Cfg
+	log         log.Logger
+	userService user.Service
+	signingKeys signingkeys.Service
+}
+
+type ExtendedJWTClaims struct {
+	jwt.Claims
+	ClientID     string              `json:"client_id"`
+	Groups       []string            `json:"groups"`
+	Email        string              `json:"email"`
+	Name         string              `json:"name"`
+	Login        string              `json:"login"`
+	Scopes       []string            `json:"scope"`
+	Entitlements map[string][]string `json:"entitlements"`
+}
+
+func (s *ExtendedJWT) Authenticate(ctx context.Context, r *authn.Request) (*authn.Identity, error) {
+	jwtToken := s.retrieveToken(r.HTTPRequest)
+
+	claims, err := s.verifyRFC9068Token(ctx, jwtToken)
+	if err != nil {
+		s.log.Error("Failed to verify JWT", "error", err)
+		return nil, errJWTInvalid.Errorf("Failed to verify JWT: %w", err)
+	}
+
+	// user:id:18
+	userID, err := strconv.ParseInt(strings.TrimPrefix(claims.Subject, fmt.Sprintf("%s:id:", authn.NamespaceUser)), 10, 64)
+	if err != nil {
+		s.log.Error("Failed to parse sub", "error", err)
+		return nil, errJWTInvalid.Errorf("Failed to parse sub: %w", err)
+	}
+
+	if r.OrgID != s.getDefaultOrgID() {
```

Comment:
Nit. Not a big deal but we could store the result of s.getDefaultOrgID() instead of calling it multiple times.
```suggestion
	// FIXME: support multiple organizations
	defaultOrgID = s.getDefaultOrgID()
	if r.OrgID != defaultOrgID {
```